libp2p/network/utils_test.go

```diff
@@ -0,0 +1,183 @@
+package network
+
+import (
+	"fmt"
+	"sort"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	"github.com/golang/mock/gomock"
+	"github.com/pokt-network/pocket/p2p/providers/addrbook_provider"
+	"github.com/pokt-network/pocket/p2p/providers/current_height_provider"
+	typesP2P "github.com/pokt-network/pocket/p2p/types"
+	mock_typesP2P "github.com/pokt-network/pocket/p2p/types/mocks"
+	"github.com/pokt-network/pocket/runtime"
+	"github.com/pokt-network/pocket/runtime/configs"
+	configTypes "github.com/pokt-network/pocket/runtime/configs/types"
+	"github.com/pokt-network/pocket/runtime/genesis"
+	coreTypes "github.com/pokt-network/pocket/shared/core/types"
+	cryptoPocket "github.com/pokt-network/pocket/shared/crypto"
+	"github.com/pokt-network/pocket/shared/modules"
+	"github.com/pokt-network/pocket/shared/modules/mocks"
+)
+
+const (
+	maxNumKeys             = 42
+	genesisConfigSeedStart = 42
+)
+
+var (
+	keys                 []cryptoPocket.PrivateKey
+	serviceUrlFormat     = "node%d.consensus:42069"
+	testServiceUrlFormat = "10.0.0.%d:42069"
+)
+
+func init() {
+	keys = generateKeys(nil, maxNumKeys)
+}
+
+func generateKeys(_ *testing.T, numValidators int) []cryptoPocket.PrivateKey {
+	keys := make([]cryptoPocket.PrivateKey, numValidators)
+
+	for i := range keys {
+		seedInt := genesisConfigSeedStart + i
+		keys[i] = cryptoPocket.GetPrivKeySeed(seedInt)
+	}
+	sort.Slice(keys, func(i, j int) bool {
+		return keys[i].Address().String() < keys[j].Address().String()
+	})
+	return keys
+}
+
+// CLEANUP: This could (should?) be a codebase-wide shared test helper
+func validatorId(i int) string {
+	return fmt.Sprintf(serviceUrlFormat, i)
+}
+
+// createMockRuntimeMgrs creates `numValidators` instances of mocked `RuntimeMgr` that are essentially
+// representing the runtime environments of the validators that we will use in our tests
+func createMockRuntimeMgrs(t *testing.T, numValidators int) []modules.RuntimeMgr {
+	ctrl := gomock.NewController(t)
+	mockRuntimeMgrs := make([]modules.RuntimeMgr, numValidators)
+	valKeys := make([]cryptoPocket.PrivateKey, numValidators)
+	copy(valKeys, keys[:numValidators])
+	mockGenesisState := createMockGenesisState(valKeys)
+	for i := range mockRuntimeMgrs {
+		cfg := &configs.Config{
+			RootDirectory: "",
+			PrivateKey:    valKeys[i].String(),
+			P2P: &configs.P2PConfig{
+				PrivateKey:     valKeys[i].String(),
+				Port:           42069,
```

Comment:
Oh I see. 

I would have done something like

`main` <- `configs` <- `libp2p-1` <- `libp2p-2` <- etc...

so that every branch has all the changes it needs and the reviewer only sees the delta because the base is the "previous" branch like you did here.

Not actionable, just thinking out loud.

Reply:
Indeed, my apologies. :sweat_smile: I wasn't quite sure what I was in for at the start and would do it differently next time.

Reply:
`...42069... 🤣 I sense a strong @Olshansk ectoplasm in this change`

It's a very impactful change